### PR TITLE
Add -acx-global-context

### PR DESCRIPTION
**Pull request by TedSander**
Add -acx-global-context so that we can rename the usages of global-context to work better in the Sass system.

### CODE REVIEW

**Comment by TedSander** [comment]
@leonsenft can you review?

**Review by leonsenft — APPROVED** [review]

**Comment by natebosch** [comment]
Do you have more context on this? I don't like the idea of leaking higher level concepts into this package.

**Comment by leonsenft** [comment]
b/182569372 for context.

Yeah, it's definitely not ideal. The history here is that we added some pseudo classes (`:host`, `:host-context()`) and combinators (`/deep/` and `>>>`) from the draft specification for Shadow DOM. At the time, Angular supported using said Shadow DOM APIs, and an emulated mode that would shim these selectors.

This package was written with hardcoded assumptions about what types of pseudo-selectors are allowed, so we had to add specific support for the Shadow DOM selectors to it.

Over time, the Shadow DOM specification dropped both `/deep/` and `>>>` (I'm not sure about the current status of `:host`), but now we had Angular relying on this package to parse them. Angular itself dropped support for Shadow DOM as well, and now relies on a custom `::ng-deep` pseudo-element in place of the `/deep/` or `>>>` combinator.

The ideal solution would be that this package supports parsing *any* pseudo-selector, and rely on other tooling to determine which pseudos are valid or not (such as a linter).

**Comment by natebosch** [comment]
Let me check my understanding:

Angular wants to start using a new non-standard psuedo-selector in CSS. This previously already happened with https://github.com/dart-lang/csslib/pull/122 and the `global-context` selector is in active use.
Angular's usage of the psuedo-selector does not match what a psuedo-selector typically means in CSS.
Sass has checks that only expected psuedo-selectors are used and will fail the build otherwise.
Sass uses `csslib` to perform those checks, and the only way to avoid the lint is to hardcode any string used as a pseudo-selector in this library.


Is that a correct summary?

Were any other designs considered for leaving parts of selectors unscoped to the angular element?
Has there been any investigation into supporting configurable allowed pseudo-selectors without hardcoding them in this library?

**Comment by natebosch** [comment]
Stepping back to https://github.com/dart-lang/csslib/pull/32 I'm confused why these pseudo-classes are hardcoded, while others like `:nth-child()` aren't...

It looks like it has to do with the stuff inside the parenthesis getting parsed differently? Why does that need to happen?

**Comment by leonsenft** [comment]
TLDR: csslib is based on a very old CSS specification; fixing this the _correct_ way would require refactoring it to support a more modern syntax (at least for pseudo-selectors).

> Angular wants to start using a new non-standard psuedo-selector in CSS. This previously already happened with dart-lang/csslib#122 and the global-context selector is in active use.

Yes. And as mentioned, this happened long before that as well when the Shadow DOM specification _dropped_ some combinators that we had added to this library.

> Angular's usage of the psuedo-selector does not match what a psuedo-selector typically means in CSS.

The `:host` selectors _do_ match the intent of their Shadow DOM equivalent: https://developer.mozilla.org/en-US/docs/Web/CSS/:host. But technically, we don't use Shadow DOM anymore so this behavior is implemented with a shim.

> Sass has checks that only expected psuedo-selectors are used and will fail the build otherwise.

It causes presubmit warnings which I believe are implemented using https://stylelint.io/.

> Sass uses csslib to perform those checks, and the only way to avoid the lint is to hardcode any string used as a pseudo-selector in this library.
>
> Is that a correct summary?

Not quite. Angular uses csslib to shim stylesheets. Neither Sass, nor Stylelint, use csslib. Stylelint by default won't lint any vendor prefixed selectors (anything with a `-` prefix), so we chose to rename `:global-context` to `:-acx-global-context` to convey that's its a custom selector and avoid the linter warning.

> Were any other designs considered for leaving parts of selectors unscoped to the angular element?

b/141198492 for background context.

> Has there been any investigation into supporting configurable allowed pseudo-selectors without hardcoding them in this library?

That would require refactoring this library to allow them.

> Stepping back to dart-lang/csslib#32 I'm confused why these pseudo-classes are hardcoded, while others like :nth-child() aren't...
> 
> It looks like it has to do with the stuff inside the parenthesis getting parsed differently? Why does that need to happen?

 The issue is that csslib was based on an older CSS specification which had hardcoded assumptions about what kinds of arguments were allowed in functional pseudo-selectors.

- `:not(<simple-selector>)`
- Anything else assumed the argument was a "selector expression" (e.g. `2n + 1`). In other words, it assumed the pseudo-selector was something like `:nth-child(2n + 1)`

[:host](https://developer.mozilla.org/en-US/docs/Web/CSS/:host()#syntax) and even [:not](https://developer.mozilla.org/en-US/docs/Web/CSS/:not#syntax) in modern CSS support different arguments. Adding support for these requires not only updating the parser, but the AST itself, which may likely be breaking.

**Comment by natebosch** [comment]
> fixing this the correct way would require refactoring it to support a more modern syntax (at least for pseudo-selectors).

Is the support you wrote for `'host'` and `'host-context'` the "modern syntax", or a separate angular specific syntax?

**Comment by leonsenft** [comment]
> Is the support you wrote for `'host'` and `'host-context'` the "modern syntax", or a separate angular specific syntax?

It's not Angular specific. It matches the grammar specification which specifies the argument is a compound selector (to the extent that csslib matches the grammar specification for parsing compound selectors, which it probably is out of date but does a decent enough job).

- https://github.com/dart-lang/csslib/blob/master/lib/parser.dart#L1559
- https://developer.mozilla.org/en-US/docs/Web/CSS/:host()#syntax

**Comment by natebosch** [comment]
After some more digging I think probably the best thing to do is keep this in for now.

@leonsenft - could you file an issue with this package including your best understanding of the gaps in this package when it comes to supporting the "modern" syntax for pseudo classes?

**Comment by leonsenft** [comment]
For sure! https://github.com/dart-lang/tools/issues/1177. As I was writing the issue I remembered the root cause is that the parser was written against the Level 3 grammar which was much more restrictive of pseudo-class function arguments. The Level 4 grammar would allow us to remove any hardcoded exceptions such as this one.
